문서 편집이력 조회 기능

kmmin78 opened the pull request:
주요 내용
1. 문서 편집이력 조회 기능을 usecase 인터페이스로 정의함으로써, 해당 프로젝트에 어떤 usecase가 있는지 보다 쉽게 파악할 수 있도록 함.
참고할만한 내용 - https://blog.cleancoder.com/uncle-bob/2011/09/30/Screaming-Architecture.html

2. 위 사항에 따른 인터페이스, 테스트코드 분리 및 생성

3. DocumentHistory 도메인 추가 (property = readonly)

4. Document의 property는 변경 가능하도록 수정

Closes #44

Commit:
문서 편집이력 조회 기능

주요 내용
1. 문서 편집이력 조회 기능을 usecase 인터페이스로 정의함으로써, 해당 프로젝트에 어떤 usecase가 있는지 보다 쉽게 파악할 수 있도록 함.
참고할만한 내용 - https://blog.cleancoder.com/uncle-bob/2011/09/30/Screaming-Architecture.html

2. 위 사항에 따른 인터페이스, 테스트코드 분리 및 생성

3. DocumentHistory 도메인 추가 (property = readonly)

4. Document의 property는 변경 가능하도록 수정

FrancescoJo commented on app-core/src/main/kotlin/kr/flab/wiki/core/domain/document/Document.kt:
```diff
@@ -6,14 +6,16 @@ import java.time.LocalDateTime
     var creator: User
 
-    val createdAt: LocalDateTime
+    var createdAt: LocalDateTime
```
계속 변경될 정보라면 `updatedAt` 라는 이름은 어떨까요?

FrancescoJo commented on app-core/src/main/kotlin/kr/flab/wiki/core/domain/document/Document.kt:
```diff
@@ -6,14 +6,16 @@ import java.time.LocalDateTime
 
     var body: String
 
     var creator: User
```
여기도, 계속 변경될 정보라면 `creator` 보다는 `lastContributor` 같은 이름은 어떨까요?

FrancescoJo commented on app-core/src/main/kotlin/kr/flab/wiki/core/domain/document/DocumentHistory.kt:
```diff
@@ -0,0 +1,33 @@
+
+@DomainModel
+interface DocumentHistory {
+    /**
```
이 코멘트는 `DocumentHistory` 타입에 대한 코멘트니까, 6라인으로 옮기는게 좋겠습니다.

FrancescoJo commented on app-core/src/main/kotlin/kr/flab/wiki/core/domain/document/DocumentHistory.kt:
```diff
@@ -0,0 +1,33 @@
+     * 근데 이러면 document의 title이 변경될 때마다 update가 필요하게 된다.
+     * 불필요한 update를 없애려면 document의 식별자는 고정된 값이 필요할 수도...?
+     */
+    val masterTitle: String
```
위에 고민하신 내용도 있습니다만 `title` 과 `masterTitle` 의 연관성을 잘 모르겠네요.
현재 구조대로라면 `title` 이 변경되면 아예 완전히 다른 문서 아닌가요?

그리고 `mastertitle` 은 불변값이라 `title` 과 무관하면서 unique 할 것 같은데요...

이런 경우 unique 한 어떤 값을 만들어주는 수단을 쓰시는 편이 괜찮을것 같아요.

참고: SecureRandom, UUID

FrancescoJo commented on app-core/src/main/kotlin/kr/flab/wiki/core/domain/document/DocumentHistory.kt:
```diff
@@ -0,0 +1,33 @@
+
+    val createdAt: LocalDateTime
+
+    val version: Long
```
version 이 불변값인가요? version 은 `Document` 에 있어야 할 속성 아닌가요?

FrancescoJo commented on app-core/src/test/kotlin/kr/flab/wiki/core/testcase/document/ShowDocumentHistoryUseCaseTest.kt:
```diff
@@ -0,0 +1,106 @@
+import org.mockito.MockitoAnnotations
+
+@Tag(TAG_TEST_UNIT)
+@DisplayName("문서 편집이력 조회 UseCase를 테스트한다.")
```
"문서 편집이력 조회 Use case 의 사양을 정의한다/확인한다" 같은 표현이 좀더 좋지 않을까 싶습니다.

FrancescoJo commented on app-core/src/test/kotlin/kr/flab/wiki/core/testcase/document/ShowDocumentHistoryUseCaseTest.kt:
```diff
@@ -0,0 +1,106 @@
+@Suppress("ClassName", "NonAsciiCharacters") // 테스트 표현을 위한 한글 사용
+class ShowDocumentHistoryUseCaseTest {
+
+    /**
```
이 주석은 좋습니다 다만, 여기가 아니라 `DocumentHistory` 파일에 `DocumentHistory` 의 도입 이유를 설명해 주시면 좋겠어요.

FrancescoJo commented on app-core/src/test/kotlin/kr/flab/wiki/core/testcase/document/ShowDocumentHistoryUseCaseTest.kt:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @Nested
+    inner class `해당 문서의 이력 조회를 했을 때` {
```
`누가` 또는 `어떤` 이 없어서 표현이 모호합니다.

FrancescoJo commented on app-core/src/test/kotlin/kr/flab/wiki/core/testcase/document/ShowDocumentHistoryUseCaseTest.kt:
```diff
@@ -0,0 +1,106 @@
+        inner class `이력이 존재하면` {
+
+            @Test
+            fun `전체 편집 이력을 조회한다`() {
```
'조회할 수 있다' 로 바꾸면 좀더 사양을 정의하는 것에 가깝게 표현할 수 있겠습니다.

FrancescoJo commented on app-core/src/test/kotlin/kr/flab/wiki/core/testcase/document/ShowDocumentHistoryUseCaseTest.kt:
```diff
@@ -0,0 +1,106 @@
+                val documentHistory = sut.findDocumentHistory(chosenDocumentTitle)
+
+                // expect:
+                assertThat(documentHistory.isEmpty(), Matchers.`is`(true))
```
`Matchers.\`is\`` 도 틀린건 아닙니다만 보통 hamcrest matcher 는 static import 로 많이 씁니다. 아래와 같은 형식으로요.

```
assertThat(documentHistory.isEmpty(), `is`(true))
```

좀더 코드가 자연스러워 지거든요.

FrancescoJo commented on app-core/src/test/kotlin/kr/flab/wiki/core/testcase/document/ShowDocumentHistoryUseCaseTest.kt:
```diff
@@ -0,0 +1,106 @@
+                // when:
+                `when`(docsRepo.findAllHistoryByTitle(chosenDocumentTitle)).thenReturn(
+                    mutableListOf(
+                        Documents.randomDocumentHistory(
```
54-71라인은 구현 중복인데요, `DOcuments.randomDocumentHistory` 메소드를 아래와 같이 바꾸면

```kotlin
object Documents {
  fun randomDocumentHistory(
    masterTitle: String,
    title: String = faker.lorem().word(),
    creator: User = Users.randomUser(),
    version: Long
  )
}
```

54-71 라인을 아래처럼 바꿀 수 있을것 같습니다.
```kotlin
  Documents.randomDocumentHistory(
    masterTitle = chosenDocumentTitle,
    version = 1
  )
  Documents.randomDocumentHistory(
    masterTitle = chosenDocumentTitle,
    version = 2
  )
  Documents.randomDocumentHistory(
    masterTitle = chosenDocumentTitle,
    version = 3
  )
```

FrancescoJo replied:
아래쪽 구현 보니 아무값 안 넣으면 이미 랜덤한 값 쓰도록 잘 구현해 두셨으니 불필요한 assignment 는 조금이라도 줄여서 테스트코드를 좀 더 짧게 바꾸면 훨씬 더 읽기 좋은 코드가 될것 같아요.

FrancescoJo commented:
테스트 충실하게 잘 만들어 주셔서 뿌듯하네요. 코멘트 한번 확인해 보시고, 좀더 고칠 부분 있나 고민해 보세요.
딱히 크게 바꿀 내용은 없으므로 이대로도 괜찮다면 저는 OK 입니다.

PR 에 적어두신 내용은 추후 README 에도 반드시 업데이트 해주세요!!

FrancescoJo reviewed (dismissed):
Minor change 를 제외한 나머지는 큰 문제 없음

neropsys commented on app-core/src/main/kotlin/kr/flab/wiki/core/domain/document/persistence/DocumentEntity.kt:
```diff
@@ -5,13 +5,13 @@ import kr.flab.wiki.core.domain.user.User
 
 internal class DocumentEntity(
-    override val title: String,
+    override var title: String,
```
전 title이 바뀐다면 전혀 다른 Document로 생각해서 val 타입으로 구현했는데, 혹시 편집이력 조회 과정에서 어떤 상황을 고려하여 title을 var로 하신것인지 궁금합니다.

kmmin78 commented:
@neropsys 저도 초기에 그렇게 생각했었는데, DocumentHistory 개념이 들어가면서 좀 혼동이 왔었네요... 문서의 최초 생성부터 title 변경도 이력으로 남겨야 된다고 생각했었는데, 말씀하신대로 전혀 다른 Document로 생각한다면 val이 맞을 것 같습니다.
수정해놓을게요 ㅎㅎ

Commit:
문서 편집이력 조회 기능

주요 내용
1. 코드리뷰 피드백 반영

2. Document title readonly로 rollback

3. Document creator -> lastContributor, createdAt -> updatedAt으로 변경

FrancescoJo commented:
@neropsys 님은 이 브랜치에서 작성하셨던 스타일로 테스트 코드를 작성해 주세요. #47 는 닫고, 테스트 올려주시면 병합하시죠.

Commit:
코드 formatter 적용

주요 내용
1. detekt에서 필터링 되는 .kt 파일 formatter 적용 (DocumentHistory 관련 파일 및 app-main, app-lib 모듈 유틸 클래스)

2. ShowDocumentHistoryUseCaseTest.kt 제거

Commit:
코드 formatter 적용

주요 내용
1. app-main 모듈의 kr.flab.wiki.app.AppConfig - MayBeConst 체크 해결
